vi/day20.md

```diff
@@ -0,0 +1,484 @@
+  event2: number;
+}
+
+@Injectable({providedIn: 'root'})
+export class SocketService {
+  private socket: Socket;
+
+  // init logic
+
+  on<EventName extends keyof SocketEventMap>(event: EventName): Observabe<SocketEventMap<EventName>> {
+    return fromEventPattern(
+      (handler) => {
+        this.socket?.on(event, handler);
+      },
+      (handler) => {
+        this.socket?.off(event, handler);
+      }
+    );
+  }
+}
+
+// usage
+this.socketService.on('event1'); // Observable<string>
+this.socketService.on('event2'); // Observable<number>
+```
+
+## `fromFetch()`
+
+`fromFetch()` is used to convert `Fetch API` to `Observable`. The usage is identical to that of `fetch()`.
+
+```ts
+// Fetch API
+fetch('https://jsonplaceholder.typicode.com/todos')
+  .then(response => response.json())
+  .then(todos => {/*...*/})
+
+// fromFetch()
+/**
+ * output:
+ * - next: [{...}, {...}]
+ * - complete
+ */
+fromFetch('https://jsonplaceholder.typicode.com/todos')
+  .pipe(
+    switchMap(response => response.json())
+  ).subscribe(observer)
+```
+
+The main difference here is that `fetch()` is `Promise` based and is eager. As soon as we invoke `fetch()`, a request will be made. `fromFetch()` converts the request to `Observable` and makes it lazy. The request **will not** be made until we call `.subscribe()` on `fromFetch()`.
+
+But wait, don't we already have `from()` for something like that? There is a _gotcha_.
+
+```ts
+from(fetch('https://jsonplaceholder.typicode.com/todos'));
+```
+
+Notice that we haven't called `.subscribe()` and the request will still be made.
+
+## `interval()`
+
+`interval()` creates an `Observable` that emits integers from `0` in a specified interval.
+
+```ts
+/**
+ * output:
+ * - (1s) next: 0
+ * - (2s) next: 1
+ * - (3s) next: 2
+ * - ...
+ */
+interval(1000).subscribe(observer);
+```
+
+`interval()` does not `complete` on its own so we will need to `unsubscribe` from `interval()` manually.
+
+## `timer()`
+
+`timer()` has two usages:
+- Create an `Observable` that will emit number `0` after a specified **delay**. This usage of `timer()` will allow it to `complete` itself.
+- Create an `Observable` that will emit integers starting with `0` after a specified **delay**, then will emit each value after a specified **interval**. This sounds similar to `interval()` but there is a slight difference which we will explore in a bit. Because this second usage is like `interval()`, it will not `complete` on its own.
+
+```ts
+/**
+ * output:
+ * - (1s) next: 0
+ * - complete
+ */
+timer(1000).subscribe(observer);
+
+/**
+ * output:
+ * - (1s) next: 0 (the first delay)
+ * - (2s) next: 1 (1s interval)
+ * - (3s) next: 2 (1s interval)
+ * - ...
+ */
+timer(1000, 1000).subscribe(observer);
+```
+
+So what's this _slight_ difference? We can pass `0` as the first argument to `timer(0, 1000)` and this effectively gives us an `Observable` that emits **right away** then every `1s` after that. `interval()` alone cannot achieve this.
```

Comment:
Just wanted to say it's a nice catch, never thought you can do something like this

Reply:
We could use this for Part 3 🙌